active/0000-element-and-fragment.md

```diff
@@ -0,0 +1,198 @@
+attributes in all derived classes. Only library authors are likely to
+need to know about them, and most users should be steered toward using
+`<element>` instead.
+
+We can also put:
+
+````js
+  attributeBindings: ['tagName', 'class']
+````
+in the base `Component` implementation so that these properties will
+propagate from the caller to the `<element>` by default.
+
+Any attributes set directly on `<element>` in a component's definition
+**take precedence over** `attributeBindings`. We make a special
+exception for `class`: `classNameBindings` and `classNames` will be
+merged with any value for `class` provided on the `<element>`. 
+
+## Fragment names
+
+Components need a dash for compatibility with web components. But this
+constraint does not apply to fragments, so fragment's names **do not**
+need to include a dash.
+
+
+## Why are `<element>` and `<fragment>` mandatory?
+
+If we want to distinguish between Component templates and Fragment
+templates, at least one of them needs to be mandatory. It makes sense
+for `<element>` to be the mandatory one, since `<element>` does useful
+work and represents a real DOM element.
+
+`<fragment>` is only needed due to upgrade-compatibility
+concerns. Some day we'll be able to treat any component without an
+`<element>` as a Fragment by default, but right now that would alter
+the behavior of existing components.
+
+Therefore, templates with neither `<element>` nor `<fragment>` will
+receive legacy behavior.
+
+## Ember.Component and Ember.Fragment classes
+
+Components must extend `Ember.Component`, the same as today.
+
+Fragments must extend `Ember.Fragment`.
+
+Mismatching the class and the template will throw an exception at
+instantiation.
+
+`Ember.Fragment` is probably an ancestor of `Ember.Component`, so that
+it can define behaviors common to both.
+
+# Drawbacks
+
+Adds a tiny amount of arguable boilerplate to templates that was not
+there before.
+
+This proposal doesn't eliminate `attributeBindings`,
+`classNameBindings`, and `classNames`. I think it would be a good idea
+to eliminate them, but we need a detailed design for some other way to
+do inheritance -- one idea would be template macro expansion.
+
+# Alternatives
+
+Instead of `<element>`, components could just use their own real tag
+in its place. But this makes it hard to distinguish template
+definition from template invocation.
+
+We could avoid introducing `<fragment>` and treat any template without
+an `<element>` as a fragment, but this makes the upgrade painful.
+
+We could detect whether a template has a single top-level element or
+not and automatically make it by a Component or Fragment. This idea
+was rejected because it's a footgun -- for example, accidentally
+changing your template from one kind to the other would change the
+sematnics of your `$` method.
+
+# Unresolved questions
+
+ - Should `Ember.Component` extend `Ember.Fragment`?
+
+ - Should `Ember.Fragment` have a `$` method that matches its DOM
```

Comment:
as its dom range is dynamic, I don't believe so. Imagine the following scenario, `this.$()` can easily become stale, and we have to mechanism by which to update it.

```hbs
<fragment>
  <h1>asdf</h1>
  {{#if theSkyIsFalling}}
    <p>Run!!</p>
  {{/if}}
</fragment>
```